File: services/api/app/schemas/charts.py
"""Chart schemas."""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


class ChartSourceResponse(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    is_mock: bool
    source_weights: dict[str, float] = Field(default_factory=dict)

    model_config = {"from_attributes": True}


class ChartEntryResponse(BaseModel):
    id: uuid.UUID
    position: int
    title: str
    artist: str
    week_date: date
    is_demo: bool
    previous_position: int | None = None
    position_change: int | None = Field(
        default=None,
        description="Positive means moved up in chart (lower rank number)",
    )

    model_config = {"from_attributes": True}


class ChartResponse(BaseModel):
    source: ChartSourceResponse
    week_date: date
    entries: list[ChartEntryResponse]


class ChartSourceWeightsUpdate(BaseModel):
    weights: dict[str, float]

File: apps/web/src/lib/charts.ts
import { getApiUrl } from "./api";

export type ChartPayload = {
  source: { name: string; slug: string; is_mock: boolean; source_weights?: Record<string, number> };
  week_date: string;
  entries: {
    position: number;
    title: string;
    artist: string;
    previous_position?: number | null;
    position_change?: number | null;
  }[];
};

export async function fetchChartClient(slug = "demo-top-40"): Promise<ChartPayload> {
  const res = await fetch(`${getApiUrl()}/api/v1/charts/${slug}`, { cache: "no-store" });
  if (!res.ok) {
    throw new Error(`API ${res.status}`);
  }
  return res.json();
}
